### PR TITLE
ElasticSearch resource lifecycle logic

### LINKED ISSUE #1775: ES Plugin: implement the plugin behaviour

### PR DESCRIPTION
**Pull request by bogdanromanx**
Fixes #1775

### CODE REVIEW

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 1]
```diff
@@ -1,21 +1,254 @@
...
 
-class ElasticSearchViews {}
+final class ElasticSearchViews private (
+    aggregate: ElasticSearchViewAggregate,
+    eventLog: EventLog[Envelope[ElasticSearchViewEvent]],
+    cache: ElasticSearchViewCache,
+    projects: Projects
+)(implicit rcr: RemoteContextResolution, uuidF: UUIDF) {
+
+  def create(
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:42]: I would not put this one as view created by other modules should follow provide a meaningful name with a pattern to be distinguishable from user-created ones

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 2]
```diff
@@ -1,21 +1,254 @@
...
+      rev: Long,
+      source: Json
+  )(implicit subject: Subject): IO[ElasticSearchViewRejection, ElasticSearchViewResource] = {
+    for {
+      p     <- projects.fetchActiveProject(project)
+      iri   <- expandIri(id, p)
+      value <- JsonLdSourceParser.decode[ElasticSearchViewValue, ElasticSearchViewRejection](p, iri, source)
+      res   <- eval(UpdateElasticSearchView(iri, project, rev, value, source, subject), p)
+    } yield res
+  }.named("updateElasticSearchView", moduleType)
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:118]: Same

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 3]
```diff
@@ -1,21 +1,254 @@
...
+      project: ProjectRef,
+      source: Json
+  )(implicit subject: Subject): IO[ElasticSearchViewRejection, ElasticSearchViewResource] = {
+    for {
+      p     <- projects.fetchActiveProject(project)
+      iri   <- expandIri(id, p)
+      value <- JsonLdSourceParser.decode[ElasticSearchViewValue, ElasticSearchViewRejection](p, iri, source)
+      res   <- eval(CreateElasticSearchView(iri, project, value, source, subject), p)
+    } yield res
+  }.named("createElasticSearchView", moduleType)
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:78]: Should the 3 create operations have different names ?

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 4]
```diff
@@ -1,21 +1,254 @@
...
+      }
+      .named("listElasticSearchViews", moduleType)
+
+  def events(offset: Offset): fs2.Stream[Task, Envelope[ElasticSearchViewEvent]] =
+    eventLog.eventsByTag(moduleType, offset)
+
+  private def currentState(project: ProjectRef, iri: Iri): IO[ElasticSearchViewRejection, ElasticSearchViewState] =
+    aggregate.state(identifier(project, iri))
+
+  private def stateAt(project: ProjectRef, iri: Iri, rev: Long) = {
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:202]: There is an operation provided by EventLogUtils that makes this a little shorter

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewSearchParams.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  * @param types      the collection of types to consider, where empty implies all views
+  * @param filter     an additional resource filter
+  */
+final case class ElasticSearchViewSearchParams(
+    project: Option[ProjectRef] = None,
+    deprecated: Option[Boolean] = None,
+    rev: Option[Long] = None,
+    createdBy: Option[Subject] = None,
+    updatedBy: Option[Subject] = None,
+    types: Set[Iri] = Set(ElasticSearch, AggregateElasticSearch).map(_.iri),
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewSearchParams.scala:28]: The way the types are matches won't work with what you have here. Either change the set here or change the `SearchParams.matches(resource)`

From:
```scala
types.subsetOf(resource.types)
```

To:
```scala
resource.types.subsetOf(types)
```

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewState.scala [hunk 6]
```diff
@@ -15,7 +15,12 @@ import io.circe.Json
 /**
   * Enumeration of ElasticSearchView state types.
   */
-trait ElasticSearchViewState extends Product with Serializable {
+sealed trait ElasticSearchViewState extends Product with Serializable {
+
+  /**
+    * @return the current resource revision
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewState.scala:21]: the current view revision

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala (hunk 3 above)
Comment (umbreak) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:78]: If that's the case, we have to change it everywhere. So far, there are no different names

Comment (imsdu) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:78]: Maybe just have two different names when we decode the json and when we are not

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala (hunk 4 above)
Comment (bogdanromanx) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:202]: it's not available... it's defined in the service

Comment (umbreak) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:202]: Maybe move it to the sdk?

Comment (bogdanromanx) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:202]: the EventLog is not available in the sdk... so the utils can't be moved.

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewBehaviours.scala [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+import monix.execution.Scheduler
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpecLike
+import org.scalatest.{Inspectors, OptionValues}
+
+import java.time.Instant
+import java.util.UUID
+import scala.concurrent.duration._
+
+trait ElasticSearchViewBehaviours {
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewBehaviours.scala:43]: There are tests missing for some of the operations in ElasticSearchViews

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewsSpec.scala [hunk 8]
```diff
@@ -1,280 +1,19 @@
...
 import ch.epfl.bluebrain.nexus.testkit.{IOFixedClock, IOValues, TestHelpers}
-import io.circe.Json
-import monix.bio.IO
-import monix.execution.Scheduler
-import org.scalatest.Inspectors
 import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpecLike
+import org.scalatest.{Inspectors, OptionValues}
 
 class ElasticSearchViewsSpec
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewsSpec.scala:8]: Why not just having 3 different suites of tests ?

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 9]
```diff
@@ -1,21 +1,398 @@
...
+  def update(
+      id: IdSegment,
+      project: ProjectRef,
+      rev: Long,
+      source: Json
+  )(implicit subject: Subject): IO[ElasticSearchViewRejection, ElasticSearchViewResource] = {
+    for {
+      p           <- projects.fetchActiveProject(project)
+      iri         <- expandIri(id, p, InvalidElasticSearchViewId.apply)
+      iriAndValue <- decode(p, Some(iri), source)
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:127]: Why this? We have introduced `betterMonadicFor`, so you should use it and do

```scala
      (_, value) <- decode(p, Some(iri), source)
```

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 10]
```diff
@@ -1,21 +1,398 @@
...
+      .named("listElasticSearchViews", moduleType)
+
+  def events(offset: Offset): fs2.Stream[Task, Envelope[ElasticSearchViewEvent]] =
+    eventLog.eventsByTag(moduleType, offset)
+
+  private def currentState(project: ProjectRef, iri: Iri): IO[ElasticSearchViewRejection, ElasticSearchViewState] =
+    aggregate.state(identifier(project, iri)).named("currentState", moduleType)
+
+  private def stateAt(project: ProjectRef, iri: Iri, rev: Long): IO[RevisionNotFound, ElasticSearchViewState] = {
+    if (rev == 0L) UIO.pure(Initial)
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:219]: I would recomment to do as in Storages and copy the `EventLogUtils` class and then use it here:
```scala
EventLogUtils
      .fetchStateAt(eventLog, persistenceId(moduleType, identifier(project, iri)), rev, Initial, next)
      .leftMap(RevisionNotFound(rev, _))
```

This will help because when porting the `EventLogUtils` to its final place we will just have to change an import instead of changing this code block

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala [hunk 11]
```diff
@@ -1,21 +1,398 @@
...
+  private def startIndexing(
+      config: ElasticSearchViewConfig,
+      eventLog: EventLog[Envelope[ElasticSearchViewEvent]],
+      index: ElasticSearchViewCache,
+      views: ElasticSearchViews
+  )(implicit as: ActorSystem[Nothing], sc: Scheduler): ActorRef[StreamSupervisor.SupervisorCommand] = {
+
+    val logger: Logger = Logger[ElasticSearchViews]
+    StreamSupervisor.runAsSingleton(
+      "ResolverIndex",
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViews.scala:373]: ElasticSearchViewIndex

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/config/AggregateConfig.scala [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+import ch.epfl.bluebrain.nexus.sourcing._
+import ch.epfl.bluebrain.nexus.sourcing.processor.{EventSourceProcessorConfig, StopStrategyConfig}
+
+/**
+  * Aggregate configuration.
+  *
+  * @param stopStrategy     the stop strategy configuration for this aggregate [[EventDefinition]]
+  * @param snapshotStrategy the snapshot strategy configuration for this aggregate [[EventDefinition]]
+  * @param processor        the event source processor config
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/config/AggregateConfig.scala:11]: Remove this. Is exists now in `ch.epfl.bluebrain.nexus.sourcing.AggregateConfig`

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/config/IndexingConfig.scala [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+import ch.epfl.bluebrain.nexus.delta.kernel.RetryStrategyConfig
+
+/**
+  * Configuration for indexing process.
+  *
+  * @param concurrency indexing concurrency
+  * @param retry       indexing retry strategy configuration
+  */
+final case class IndexingConfig(concurrency: Int, retry: RetryStrategyConfig)
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/config/IndexingConfig.scala:11]: Remove this. Is exists now in `ch.epfl.bluebrain.nexus.delta.kernel. IndexingConfig `

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewRejection.scala [hunk 14]
```diff
@@ -32,21 +36,24 @@ object ElasticSearchViewRejection {
     *
     * @param tag the provided tag
     */
-  final case class TagNotFound(tag: Label) extends ElasticSearchViewRejection(s"Tag requested '$tag' not found.")
+  final case class TagNotFound(tag: Label, id: Iri, project: ProjectRef)
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewRejection.scala:39]: Let's keep consistency. All the other resource types have the following fields
```scala
  final case class TagNotFound(tag: Label) extends ResolverRejection(s"Tag requested '$tag' not found.")
```

So either you keep it here in the same way or change it in all the others.

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewState.scala [hunk 15]
```diff
@@ -106,7 +113,7 @@ object ElasticSearchViewState {
           id = id,
           uris = ResourceUris("views", project, id)(mappings, base),
           rev = rev,
-          types = Set(value.tpe.iri),
+          types = Set(value.tpe.iri, nxv.View),
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewState.scala:116]: Isn't it better to have the type logic in `ElasticSearchType`?

```scala
sealed ElasticSearchType {
  def iri: Iri
    def types: Set[Iri] = Set(nxv.View, iri)

 }
```

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/AbstractDbSpec.scala [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+import org.scalatest.wordspec.AnyWordSpecLike
+import org.scalatest.{BeforeAndAfterAll, CancelAfterFailure}
+import slick.jdbc.H2Profile.api._
+import slick.jdbc.JdbcBackend
+import slick.jdbc.JdbcBackend.Database
+
+import java.util.UUID
+import scala.concurrent.duration._
+
+abstract class AbstractDbSpec
```
Comment (umbreak) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/AbstractDbSpec.scala:19]: Add a `TODO` for this to be moved somewhere else. We don't want to duplicate this class for each module

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewValue.scala [hunk 17]
```diff
@@ -33,14 +41,14 @@ object ElasticSearchViewValue {
...
-      includeDeprecated: Boolean,
+      resourceSchemas: Set[Iri] = Set.empty,
+      resourceTypes: Set[Iri] = Set.empty,
+      resourceTag: Option[Label] = None,
+      sourceAsText: Boolean = true,
+      includeMetadata: Boolean = true,
+      includeDeprecated: Boolean = true,
       mapping: Json,
-      permission: Permission
+      permission: Permission = Permission.unsafe("views/query")
```
Comment (wwajerowicz) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewValue.scala:51]: The default value should either come from config or be defined as a constant somewhere.

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewDecodingBehaviours.scala [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+import ch.epfl.bluebrain.nexus.delta.sdk.utils.UUIDF
+import ch.epfl.bluebrain.nexus.testkit.{IOValues, TestHelpers}
+import io.circe.literal._
+import org.scalatest.Inspectors
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpecLike
+
+import java.util.UUID
+
+trait ElasticSearchViewDecodingBehaviours {
```
Comment (wwajerowicz) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewDecodingBehaviours.scala:21]: Couldn't this just be a regular test class?

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewSTMBehaviours.scala [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+import monix.bio.IO
+import monix.execution.Scheduler
+import org.scalatest.Inspectors
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpecLike
+
+import java.time.Instant
+import java.util.UUID
+
+trait ElasticSearchViewSTMBehaviours {
```
Comment (wwajerowicz) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/ElasticSearchViewSTMBehaviours.scala:29]: Couldn't this just be a regular test class?

**Review by wwajerowicz — CHANGES_REQUESTED** [review]

**Comment by wwajerowicz** [comment]
General comment: wouldn't it make sense to rename all the commands to have shorter names e.g. `CreateElasticSearchView`=> `Create` since they are nested inside `ElasticSearchViewCommand` and probably won't be used anywhere outside of the plugin? same with events?

**Comment by umbreak** [comment]
I thing the naming is ok. Otherwise when searching for `Create` we are gonna have many collisions, and that's probably more annoying than making the names smaller

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewValue.scala (hunk 17 above)
Comment (umbreak) [reply delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/model/ElasticSearchViewValue.scala:51]: We don't have them as configuration values (as we don't have them for anything else. But I agree it should be a constant somewhere, for visibility

File: delta/plugins/elasticsearch/src/test/resources/akka-persistence-test.conf [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+akka-kryo-serialization {
+  id-strategy = "automatic"
+  implicit-registration-logging = true
+  resolve-subclasses = false
+}
+
+h2 {
+  profile = "slick.jdbc.H2Profile$"
+  db {
+    url = "jdbc:h2:mem:"${test-database}";DATABASE_TO_UPPER=false;"
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/test/resources/akka-persistence-test.conf:32]: I changed test-database to test-instance in my last PR to create a custom cache directory

**Review by wwajerowicz — APPROVED** [review]

File: docs/src/main/paradox/docs/faq.md [hunk 21]
```diff
@@ -121,8 +121,7 @@ exploration of trends and patterns.
 
 @link:[SHACL](https://www.w3.org/TR/shacl/){ open=new } (Shapes Constraint Language) is a language for validating RDF 
 graphs against a set of conditions. These conditions are provided as shapes and other constructs expressed in the form 
-of an RDF graph. SHACL is used in Blue Brain Nexus to constrain and control the payload that can be pushed into Nexus. 
-You can use the @link:[SHACL Playground](https://shacl.org/playground/){ open=new } to test your schemas.
+of an RDF graph. SHACL is used in Blue Brain Nexus to constrain and control the payload that can be pushed into Nexus.
```
Comment (umbreak) [inline docs/src/main/paradox/docs/faq.md:124]: Revert this change, since they have already fixed it https://github.com/TopQuadrant/shacl-js/issues/29

**Review by umbreak — CHANGES_REQUESTED** [review]

File: docs/src/main/paradox/docs/faq.md (hunk 21 above)
Comment (bogdanromanx) [reply docs/src/main/paradox/docs/faq.md:124]: thanks for checking! ;)

**Review by imsdu — APPROVED** [review]

**Review by umbreak — APPROVED** [review]
